### PR TITLE
babel-generator: Add TypeScript support

### PR DESCRIPTION
**Pull request by andy-ms**
| Q                        | A
| ------------------------ | ---
| Patch: Bug Fix?          |  no
| Major: Breaking Change?  | no
| Minor: New Feature?      | yes
| Deprecations?            | no
| Spec Compliancy?         | no
| Tests Added/Pass?        | tests pass
| Fixed Tickets            |
| License                  | MIT
| Doc PR                   |
| Dependency Changes       | no

Adds generators for the new TypeScript-specific nodes and for new modifiers from TypeScript (such as `abstract`).
Would prefer to get #5856 in first and then merge that in to this. They are mostly independent, but see comments in `parentheses.js` and `test/index.js`.

### CODE REVIEW

File: packages/babel-generator/test/index.js [hunk 1]
```diff
@@ -367,7 +368,7 @@ suites.forEach(function(testSuite) {
             if (actualCode) {
               const actualAst = parse(actualCode, {
                 filename: actual.loc,
-                plugins: [
+                plugins: task.options.plugins || [
```
Comment (hzoo) [inline packages/babel-generator/test/index.js]: maybe it's better we be explicit in the tests 👍 , could probably remove flow (or all of them) and add a plugin option for each test later.

File: packages/babel-generator/test/generate-typescript-tests.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
+Copies tests from babylon's TypeScript test suite to babel-generator.
```
Comment (hzoo) [inline packages/babel-generator/test/generate-typescript-tests.js]: kinda thinking that is another reason why we should move babylon back into the monorepo?

File: packages/babel-generator/test/index.js (hunk 1 above)
Comment (andy-ms) [inline packages/babel-generator/test/index.js]: Added #6018.

**Comment by andy-ms** [comment]
I forgot to ask, when is it appropriate to call `this.token("foo")` vs `this.word("foo")`?

**Comment by hzoo** [comment]
@andy-ms I'm not sure myself but looking at what we have it looks like tokens are more like "symbols" like `+`, `*` while word is actually some characters like a keyword?

**Pull request merged**